# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/flyrank-bih/flyrank-ml-internship-starter/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

**Lane:** CTR / Engagement Opportunity Scoring

**Why:** High-impression, low-CTR pages represent the highest ROI opportunity for any content or growth team. Instead of blindly guessing which pages to update, this lane allows us to systematically identify "near-miss" pages. By modeling expected CTR against actual CTR, adjusted for SERP position and search volume, we can prioritize editorial efforts exactly where they will yield the greatest immediate traffic lift. This moves content strategy from reactive to highly data-driven.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

* **The Decision:** Which existing, high-visibility pages should the editorial team prioritize for meta-title, snippet, or above-the-fold content optimization?
* **The Action:** Rewriting title tags, meta descriptions, or restructuring the introduction for the top-ranked pages in our opportunity queue.
* **The Unit of Analysis:** A single web page (`content_id`).
* **Cost of a Wrong Call (False Positive):** Wasting expensive editorial hours tweaking a page that has low true growth potential or is already performing at its baseline maximum.
* **Cost of a Missed Opportunity (False Negative):** Leaving high-volume, easily captured traffic on the table while competitors out-click us on the exact same search engine result pages.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [3]:
import pandas as pd
df = pd.read_csv('/content/content_refresh_anonymized.csv')

# This will print out all the exact column names in your dataset
print(df.columns.tolist())

['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [4]:
import pandas as pd

# Load the anonymized starter dataset from the repository
df = pd.read_csv('/content/content_refresh_anonymized.csv')

# 1. Scope: Total pages we are analyzing
total_pages = len(df)

# 2. Opportunity: Pages with high visibility but poor click-through
high_imp_low_ctr = len(df[(df['impressions_90d'] >= 1000) & (df['ctr'] < 0.02)])

# 3. Context: What is the average position of these underperforming pages?
avg_pos = df[(df['impressions_90d'] >= 1000) & (df['ctr'] < 0.02)]['avg_position'].mean()

print(f"Dataset scope: {total_pages:,} total pages.")
print(f"Immediate Opportunity Size: {high_imp_low_ctr:,} pages have >1,000 recent impressions but a CTR under 2%.")
print(f"Average rank of these missed opportunities: Position {avg_pos:.1f}.")

Dataset scope: 30,000 total pages.
Immediate Opportunity Size: 1,235 pages have >1,000 recent impressions but a CTR under 2%.
Average rank of these missed opportunities: Position 25.2.


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

* **What I can claim:**
  * **Observed Performance Gaps:** In our 30,000-page dataset, we identified 1,235 high-visibility pages (>1,000 recent impressions) with underperforming CTRs (<2%) sitting at an average position of 25.2.
  * **Directional Decision-Support:** This work provides a position- and volume-adjusted opportunity queue to help editorial teams focus resources where the gap between expected and actual performance is largest.
  * **Systematic Benchmarking:** We can establish baseline expected CTR curves across position tiers to measure relative content performance objectively.

* **What I CANNOT claim:**
  * **Causal Guarantees:** This observational model cannot guarantee that updating a title tag or snippet will cause a specific percentage increase in traffic (verifying true causal impact requires controlled A/B experimentation).
  * **Algorithm Reverse-Engineering:** This model does not "predict Google's algorithm" or account for unobserved SERP layout dynamics (e.g., featured snippets, local packs, or ad placement variations).

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.